tests/drivers/uart/uart_pm/src/main.c

```diff
@@ -0,0 +1,265 @@
+		 * not hang.
+		 */
+		for (int i = 0; i < ARRAY_SIZE(outs); i++) {
+			uart_poll_out(dev, outs[i]);
+		}
+
+		return;
+	}
+
+	err = uart_poll_in(dev, &c);
+	zassert_equal(err, -1, NULL);
+
+	for (int i = 0; i < ARRAY_SIZE(outs); i++) {
+		uart_poll_out(dev, outs[i]);
+		k_busy_wait(1000);
+
+		if (active) {
+			err = uart_poll_in(dev, &c);
+			zassert_equal(err, 0, "Unexpected err: %d", err);
+			zassert_equal(c, outs[i], NULL);
+		}
+
+		err = uart_poll_in(dev, &c);
+		zassert_equal(err, -1, NULL);
+	}
+}
+
+static void async_callback(const struct device *dev, struct uart_event *evt, void *ctx)
+{
+	bool *done = ctx;
+
+	switch (evt->type) {
+	case UART_TX_DONE:
+		*done = true;
+		break;
+	default:
+		break;
+	}
+}
+
+static bool async_verify(const struct device *dev, bool active)
+{
+	char txbuf[] = "test";
+	uint8_t rxbuf[32];
+	volatile bool tx_done = false;
+	int err;
+
+	err = uart_callback_set(dev, async_callback, (void *)&tx_done);
+	if (err == -ENOTSUP) {
+		return false;
+	}
+
+	if (!active) {
+		return true;
+	}
+
+	zassert_equal(err, 0, "Unexpected err: %d", err);
+
+	if (HAS_RX) {
+		err = uart_rx_enable(dev, rxbuf, sizeof(rxbuf), 1);
+		zassert_equal(err, 0, "Unexpected err: %d", err);
+	}
+
+	err = uart_tx(dev, txbuf, sizeof(txbuf), 10);
+	zassert_equal(err, 0, "Unexpected err: %d", err);
+
+	k_busy_wait(10000);
+
+	if (HAS_RX) {
+		err = uart_rx_disable(dev);
+		zassert_equal(err, 0, "Unexpected err: %d", err);
+
+		k_busy_wait(10000);
+
+		err = memcmp(txbuf, rxbuf, sizeof(txbuf));
+		zassert_equal(err, 0, "Unexpected err: %d", err);
+	}
+
+	zassert_true(tx_done, NULL);
+
+	return true;
+}
+
+static void communication_verify(const struct device *dev, bool active)
+{
+	bool is_async = async_verify(dev, active);
+
+	polling_verify(dev, is_async, active);
+}
+
+#define state_verify(dev, exp_state) do {\
+	uint32_t power_state; \
+	int err = pm_device_state_get(dev, &power_state); \
+	zassert_equal(err, 0, "Unexpected err: %d", err); \
+	zassert_equal(power_state, exp_state, NULL); \
+} while (0)
+
+static void pm_callback(const struct device *dev,
+			int status, uint32_t *state, void *arg)
+{
+	zassert_equal(dev, exp_dev, NULL);
+	zassert_equal(status, 0, NULL);
+	zassert_equal(*state, exp_state, NULL);
+	zassert_equal(arg, exp_arg, NULL);
+	callback_cnt++;
+}
+
+static void state_set(const struct device *dev, uint32_t state, int exp_err, bool cb)
+{
+	int err;
+	uint32_t prev_state;
+
+	err = pm_device_state_get(dev, &prev_state);
+	zassert_equal(err, 0, "Unexpected err: %d", err);
+
+	if (cb) {
+		callback_cnt = 0;
+		exp_dev = dev;
+		exp_arg = &state;
+		exp_state = state;
+
+		err = pm_device_state_set(dev, state, pm_callback, exp_arg);
+		zassert_equal(err, exp_err, "Unexpected err: %d", err);
+		zassert_equal(callback_cnt, 1, NULL);
+	} else {
+		err = pm_device_state_set(dev, state, NULL, NULL);
+		zassert_equal(err, exp_err, "Unexpected err: %d", err);
+	}
+
+	state_verify(dev, err == 0 ? state : prev_state);
```

Comment:
I think you need a local variable for `err == 0 ? state : prev_state`, as the macro uses its own local `err` variable, and in the current form the `exp_state` parameter will be just replaced with the above expression, so in effect that local `err` will be checked instead of the outer one.